src/System.Drawing.Common/src/Configurations.props

```diff
@@ -1,11 +1,16 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <BuildConfigurations>
+    <PackageConfigurations>
       netcoreapp2.0-Windows_NT;
       netcoreapp2.0-Unix;
       netfx;
       netstandard;
+    </PackageConfigurations>
+    <BuildConfigurations>
```

Comment:
This is turning System.Drawing.Common into yet another partial OOB, with unmanaged dependency. Partial OOBs like that have very high maintenance costs. If we want to do this, it should be approved by the .NET Core disro owners.
cc @Petermarcu @karelz 

The right way to solve the problem with PInvoking to DlOpen is to add the imperative dll import APIs to .NET Core 2.1: https://github.com/dotnet/coreclr/issues/14968. It will avoid the need for the dependency on .NE T Core shims, and putting System.Drawing.Common inbox.

Reply:
@jkotas, won't it be a partial OOB either way? A build for 2.0 on top of standard, and then a build for 2.1 using netcoreapp-specific APIs. How does a dependency on a System.Native.so/dylib function cause any more issues than a dependency on netcoreapp-specific method?

Reply:
The difference is whether it is part of .NET Runtime download or whether the app carries it with itself.

Reply:
I don't think dotnet/coreclr#14968 will solve the problem this PR is trying to address.

Yes, there are a lot of cases where managed code wants to have a lot of control over which library is being loaded. Managed libraries which wrap around native libraries, for example, will know where their native library is located, what name(s) it may have, and will try to load that directly.
For example, I think it's normal System.Drawing.Common has a very strong opinion about where libgdiplus.so is located, because it has domain knowledge about libgdiplus. The same for, say,  FFmpeg.AutoGen, which is a managed wrapper around  ffmpeg.

On the other hand, System.Drawing.Common doesn't care at all about dlopen and has no domain knowledge about it. The way I understand it, dotnet/coreclr#14968 would allow System.Drawing.Common to say something along the lines of "dlopen may be in libdl.so on most Linux distro's, libdl.so.2 on CentOS and something else on FreeBSD".

I don't think it's the business of System.Drawing.Common to care about that.

Anyone writing a managed wrapper around a native library trying and using function pointers will need dlopen and dlsym. Everyone doing that will need to replicate the same search logic.
And most importantly, that logic will eventually fail: when corefx is ported to a new platform, when libdl versions,... .

Long story short, for this use case, I believe a better option would be to make dlopen/dlsym in one shape or another part of the CoreFX API, as suggested in #17135.

"The least that could possible work" may be to expose a public `LoadNativeLibrary` and `LoadNativeSymbol` API in CoreFX on which System.Drawing.Common and others can depend.

Reply:
> I believe a better option would be to make dlopen/dlsym in one shape or another part of the CoreFX API, as suggested in #17135.

Agree. I did not know that #17135 existed. The APIs proposed by #17135 are the kind of APIs that I thought we will add as part of dotnet/coreclr#14968.

Reply:
> The difference is whether it is part of .NET Runtime download or whether the app carries it with itself. Everything that depends on System.Native.so/dylib shim should be part of .NET Runtime download. The System.Native.so/dylib shim is not part of our public API surface.

That's a new requirement.  It might be the right one, but I don't believe we've explicitly stated that before, nor do I see a technical limitation that would require it.  The native shims are considered internal surface area to coreclr/corefx, but this System.Drawing.Common component is part of corefx, regardless of what package it's delivered in.  We could of course tighten the restriction to the requirement you've listed, which has its own benefits but of course also its own negatives.  And for better or worse, we already have some such dependencies, e.g. System.Data.SqlClient isn't part of the .NET Runtime download but it does P/Invoke into the System.Net.Security.Native.so/dylib shim.

Reply:
> The native shims are considered internal surface area to coreclr/corefx, but this System.Drawing.Common component is part of corefx, regardless of what package it's delivered in.

This works only if the package bundles local private copy of the shim, or if the package declares dependency on the shims and we maintain and test the shims as public APIs. It is not the case today - e.g. if you take all .NET Core 2.0 shims and copy them over to .NET Core 1.0, it is not going to work.

Logically, we should think about each of the indepedently shipped components as liviving in its own repo.


> And for better or worse, we already have some such dependencies, e.g. System.Data.SqlClient isn't part of the .NET Runtime download but it does P/Invoke into the System.Net.Security.Native.so/dylib shim.

We either need to fix that, or we need to add a test coverage to ensure that you can mix and match different versions of .NET Core Runtime (that bundles `System.Net.Security.Native.so/dylib`) with different versions of `System.Data.SqlClient` and document/enforce the supported mixing and matching.

The packaging for `System.Data.SqlClient` is suspect in general: All build configurations in https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/src/Configurations.props are netstandard. How am I going to get the .NET Core shims that the implementation depends on when I install it on Mono?

Opened https://github.com/dotnet/corefx/issues/25205

Reply:
> This works only if the package bundles local private copy of the shim

I'm not following.  The shims are part of the runtime, are they not?  So if an assembly is built for netcoreapp2.1, and netcoreapp2.1's runtime includes the shims, what fails?

> if you take all .NET Core 2.0 shims and copy them over to .NET Core 1.0, it is not going to work

Sure, but that's not the case we're talking about here.  This assembly was declaring that it required netcoreapp2.1.

Reply:
> or we need to add a test coverage to ensure that you can mix and match different versions of .NET Core Runtime

Right, there are strings attached either way.

Reply:
If nuget package declares dependency on netcore2.1, it is expected to work on netcoreapp2.2 or even netcoreapp3.0 as well. How do we make sure that it works there when it depends on netcoreapp2.1 shims?

Reply:
> How do we make sure that it works there when it depends on netcoreapp2.1 shims?

By taking such a dependency we're signing up to keep those functions exposed, as part of the internal surface area of the product, little different than signing up to continue keeping public surface area exposed in subsequent versions.

Like I said earlier, I'm fine if we decide against that, and want to impose the restrictions you highlight.  I'm simply calling out that I don't think it's the only avenue available to us.

Reply:
Agree.

The difference is that we have a lot of infrastructure to maintain compatibility of the public surface. Such infrastructure does not exist for the internal surface. It would need to be built. 

We have tried to play similar internal surface games in .NET Framework in the past (e.g. with `System.SizedReference`). It tends to turn into poorly designed, tested and documented part that everybody is afraid to touch. I would love to keep things simple and stick to public surface when it comes to dependencies between independent packages.

Reply:
> The difference is that we have a lot of infrastructure to maintain compatibility of the public surface.

Fair enough.

> I would love to keep things simple and stick to public surface when it comes to dependencies between independent packages.

Ok.

Reply:
FWIW I agree with @jkotas that we should avoid taking these type of dependencies in independent shipping packages. While the rule likely isn't written down anywhere we do try to follow the same strategy for anything the depends on System.Private.CoreLib directly as well.